// containerful/dist/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var appdata_path_1 = __importDefault(require("appdata-path"));
var path_1 = __importDefault(require("path"));
exports.GRAPHQL_URL = 'http://localhost:8080/graphql';
exports.DOCKER_COMPOSE_CONFIG_FIELD = 'x-containerful';
exports.DEFAULT_REGION = 'us_central1';
exports.LOCALHOST_LISTENER = '/onLogin';
exports.LOGIN_STATIC_SITE_DIR = 'loginStaticSite';
exports.NEXT_SITE_URL = 'http://localhost:3000';
exports.USER_TOKEN_CONFIG_KEY = 'token';
var APP_DATA_FOLDER = 'containerful';
exports.CONFIGSTORE_NAME = 'config';
exports.APP_DATA_PATH = path_1.default.resolve(appdata_path_1.default(APP_DATA_FOLDER));
exports.FIREBASE_CONFIG = {
    apiKey: 'AIzaSyCc-zUlvJ5gnISC9I-FU9hKlAbfXoevxoM',
    authDomain: 'containerful-15f56.firebaseapp.com',
    databaseURL: 'https://containerful-15f56.firebaseio.com',
    projectId: 'containerful-15f56',
    storageBucket: 'containerful-15f56.appspot.com',
    messagingSenderId: '799298051038',
    appId: '1:799298051038:web:afd3d76d0db32fbc82db4f',
    measurementId: 'G-ZZ8G3ZVSQ8',
};
